library/core/src/mem/transmutability.rs

```diff
@@ -13,6 +66,53 @@ pub unsafe trait BikeshedIntrinsicFrom<Src, const ASSUME: Assume = { Assume::NOT
 where
     Src: ?Sized,
 {
+    /// Transmutes a `Src` value into a `Self`.
+    ///
+    /// # Safety
+    ///
+    /// The safety obligations of the caller depend on the value of `ASSUME`:
+    /// - If [`ASSUME.alignment`](Assume::alignment), the caller must prove that
+    ///   all references in `src` (including `src` itself, if `Src` is a
+    ///   reference type) have well-aligned referents upon transmutation into
+    ///   `Self`.
+    /// - If [`ASSUME.lifetimes`](Assume::lifetimes), the caller must prove that
+    ///   all references in `src` (including `src` itself, if `Src` is a
+    ///   reference type) have referents that live as long as `Self`.
```

Comment:
Same comment here about referring to objects in `src` vs in the resulting `Self`. However, this gets trickier since the rules for what constitute valid lifetimes are super tricky. My first inclination was to punt and say something that just passes the buck to Rust, e.g.:

> the caller must prove that it is sound to transmute `src` into `Self` with the given lifetimes

...but of course, that demands that the caller prove the soundness of the *entire* transmutation, not just the lifetime aspects of the transmutation. I'm honestly not sure how you'd express the idea of "the lifetime aspects of a transmutation" in precise language.

At a minimum, any such language would need to refer to the referents of references since references are valid for particular lifetimes *in virtue of* what happens with their referents in other parts of the program. I'm thinking of something similar to the language we use regarding reference validity in zerocopy's `Ptr` type, where we talk a lot about the referent.

Reply:
Hm, yeah, this needs substantial reworking. The intention of this `Assume` switch is to permit using this trait/method for lifetime extension. A common case of this is extending lifetimes to `'static`, even though such lifetimes rarely accurately describe the referent.

I think your inclination to phrase this obligation with respect to what happens *after* the transmutation is correct. (And a similar complaint can be levied against `Assume::safety`.)

Reply:
@joshlf, what do you think of:
```suggestion
    /// - If [`ASSUME.lifetimes`](Assume::lifetimes), the caller must guarantee
    ///   that pointers in the returned `Self` will not dangle.
```
This is intentionally conservative, allowing for the language validity rules of references to be relaxed in the future. For instance, it may eventually be decided that non-dangling-ness is a library invariant, rather than a language invariant. If that happens, we can relax this obligation to "the caller must guarantee that no dangling dereferences occur through the returned `Self`" without breaking any existing proofs.

Reply:
I like that, although two tiny nits:
- Instead of "pointers", I would say "references"
- Instead of "will not dangle", I would say "will not outlive their referents"

Reply:
[Done](https://github.com/rust-lang/rust/compare/96cfefbb125916871760c1df4c2ee29944635e90..1c0e7fac16490ea2e83ab0127c364231a42742ad#diff-85698661a7eb53801df364c9242c8613f7d2e4bca026bc5c88d542c22fa01ddfR90-R92)